### ISSUE TITLE
[ES6 support] Cannot reflectively construct classes in ES6 mode

### ISSUE THREAD

**Issue by alexeagle** [angular/angular#7740]
Because we reflectively create instances of things defined as ES6 class and constructor, we get a Chrome error:
`constructor cannot be invoked without 'new'`

**Comment by stringham** [comment]
We are running into this issue. Are there any known workarounds?

**Comment by alexeagle** [comment]
no, I didn't find a workaround. We have not been running any ES6 code in the browser so we haven't run across this since I reported it. I'm not sure where the reflection occurs, do you have a stack trace handy?
We may need @mhevery to assist in how we reflectively instantiate things.

**Comment by vicb** [comment]
https://github.com/angular/angular/blob/190bcc89c115ef1d0db898c3ced032dac25d8ae3/modules/%40angular/core/src/util/decorators.ts#L290 ?

**Comment by alexeagle** [comment]
yes, that looks right. Do you know what the fix should be?

**Comment by vicb** [comment]
Yes, I know

**Comment by stringham** [comment]
I believe `new annotationCls(...args)` works in this place, but the problem also happens here:

A runtime compiled file: `_NoopModule.ngfactory.js`

```function _NoopModuleInjector(parent) {
  var self = this;
  jit_AppModuleInjector0.call(this, parent,[]); // <-- calling a class without 'new'
}```

**Comment by vicb** [comment]
The factories are transpiled, the same should work

**Comment by vicb** [comment]
Wait because they are transpiled it should not be a pb, is it ? (the class should not use `class`)

**Comment by alexeagle** [comment]
Ryan is trying ES6 at runtime, so there are classes, and the generated code
would need to allow this as well.

**Comment by tbosch** [comment]
I think we should then emit and eval ES6 code... Needs a new emitter and a Config flag

**Comment by imcotton** [comment]
Also find out that [`ReflectionCapabilities`][reflection_capabilities] has the same issue, I think having ES6 class compatibility would be much relief during development stage from daily bases. Maybe consider using [Reflect.construct] instead?


[Reflect.construct]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect/construct
[reflection_capabilities]: https://github.com/angular/angular/blob/292ccf882ae2d3f0f816d81d3a19edff753393f7/modules/%40angular/core/src/reflection/reflection_capabilities.ts#L186

**Comment by irwansyahwii** [comment]
Hi guys! I just updated my project using Angular 2 final release but I then I experienced this issue since on the server side I use Angular 2 DI independently. @vicb would you please told me what the solution for this? :(

**Comment by irwansyahwii** [comment]
FYI guys, Angular 2 RC5 works with ES6 classes so I wonder why the final release doesn't use the RC5 code?

**Comment by stofte** [comment]
The easiest fix for me was to target es5 from TypeScript instead of es2015, and include core-js and live with it (since I'm just consuming the npm umd bundles). 

This was also working in RC1, from which I upgraded.

**Comment by zakhenry** [comment]
Same issue here - targeting es2015 in tsconfig as I'm using the DI as a component for a node backend server.

**Comment by calebboyd** [comment]
@stofte That'd be a great solution, But Typescript doesn't support es6 class features (correct inheritance etc) in when targeting es5 :(

**Comment by calebboyd** [comment]
fixed in #12099
